### PR TITLE
feat: path handling UX improvements

### PR DESCRIPTION
**Pull request by tustvold**
A couple of things came up whilst implementing https://github.com/apache/arrow-datafusion/pull/2677 that would make life easier.

* `Path::parse("/")` would return an error
* No easy way to convert a `std::path::Path` to `object_store::path::Path` for `LocalFileSystem`
* Prefix match was private

This PR addresses these

### CODE REVIEW

File: src/path/mod.rs [hunk 1]
```diff
@@ -233,6 +302,22 @@ mod tests {
         assert_eq!(location.as_ref(), "foo%2Fbar/baz%252Ftest");
     }
 
+    #[test]
+    fn test_parse() {
+        assert_eq!(Path::parse("/").unwrap().as_ref(), "");
+        assert_eq!(Path::parse("").unwrap().as_ref(), "");
+
+        let err = Path::parse("//").unwrap_err();
```
Comment (alamb) [inline src/path/mod.rs:310]: 👍

File: src/path/mod.rs [hunk 2]
```diff
@@ -214,6 +271,18 @@ where
     }
 }
 
+pub(crate) fn filesystem_path_to_url(path: impl AsRef<std::path::Path>) -> Result<Url, Error> {
```
Comment (alamb) [inline src/path/mod.rs:274]: Perhaps a docstring explaining what this does would be useful as well

File: src/path/mod.rs [hunk 3]
```diff
@@ -120,6 +144,44 @@ impl Path {
...
+    /// This will return an error if the path does not exist, or contains illegal
+    /// character sequences as defined by [`Path::parse`]
+    pub fn from_filesystem_path(path: impl AsRef<std::path::Path>) -> Result<Self, Error> {
+        Self::from_filesystem_path_with_base(path, None)
+    }
+
+    /// Convert a filesystem path to a [`Path`] relative to the provided base
+    ///
+    /// This will return an error if the path does not exist, contains illegal
+    /// character sequences as defined by [`Path::parse`], or does not start
```
Comment (alamb) [inline src/path/mod.rs:158]: I found "provided prefix" confusing as I didn't know what it was referring to. It seems to  mean that the `path` must be prefixed by the the path in the `base` URL, if specified

**Review by alamb — APPROVED** [review]
LGTM -- I had some suggestions on docstrings but otherwise 👌

File: src/path/mod.rs (hunk 3 above)
Comment (tustvold) [reply src/path/mod.rs:158]: Oh sorry, I renamed the parameters and forgot to update the docstring

**Pull request merged**